src/main/java/net/minecraftforge/common/util/ChunkedTileEntityList.java

```
@@ -0,0 +1,283 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2017.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+package net.minecraftforge.common.util;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.Set;
+
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.math.ChunkPos;
+import net.minecraft.world.World;
+
+public class ChunkedTileEntityList implements Collection<TileEntity>
```

Shouldn't this class implement `List`?

---

No, because implementing methods like `get(int)` would be horrible performance wise. People should be forced to use the iterator.

---

Maybe it is better to just deprecate those methods and throw a unsupported operation exception? Right now you just change the field type, while devs may still expect a `List`.

---

Already thought about that one, but it wouldn't make a difference in my most cases. It would only change the exception which will be thrown once somebody uses those methods (like get(int), remove(int)).

The only case where it would make a difference would be if a mod casts it to a list (maybe to store it somewhere else). That is rather unlikely, but not impossible. So I'm not really sure about it.

I will leave it for now as it is, to keep it better organized. Maybe someone else knows if this is necessary or not.

---

I think a `List` is more preferred as it maintains (binary) compatibility. Anything that is declared as public might be used by some mod, so best to play safe and let it be a List.

---

Making it a `List` will maintain *source* compatibility, I think. Breaking binary compatibility is not too bad, since mods can just recompile against the new version, but breaking compatibility with existing code like this currently does is not an option, I'd say.

---

Ok, looks like java is a little bit weird and refers to the list interface instead of the parent one. I will change it to a list real quick.